src/gui/accountstate.cpp

```diff
@@ -305,12 +306,17 @@ void AccountState::slotInvalidCredentials()
     qCInfo(lcAccountState) << "Invalid credentials for" << _account->url().toString()
                            << "asking user";
 
-    if (account()->credentials()->ready())
+    _waitingForNewCredentials = true;
+    setState(AskingCredentials);
+
+    if (account()->credentials()->ready()) {
         account()->credentials()->invalidateToken();
+        if (auto creds = qobject_cast<HttpCredentials *>(account()->credentials())) {
+            if (creds->refreshAccessToken())
+                return;
```

Comment:
Yes, i know, it breaks the abstraction layer. 
I tought about putting refreshAccessToken in AbstractCredentials.  But i think in the long term we could get rid of it and only have HttpCredentials.
If you insist i can do it.

(Note: there are other places where i broke this abstraction layer)